File: Java/src/_0671_Second_Minimum_Node_In_a_Binary_Tree/Solution.java
package _0671_Second_Minimum_Node_In_a_Binary_Tree;

/**
 * Created by WXX on 2021/3/4 14:45
 * 执行用时：0 ms, 在所有 Java 提交中击败了100.00%的用户
 * 内存消耗：35.6 MB, 在所有 Java 提交中击败了76.32%的用户
 */
public class Solution {

    long d1 = Long.MAX_VALUE, d2 = Long.MAX_VALUE;

    private void dfs(TreeNode root) {
        if (root == null) return;
        int x = root.val;
        if (x < d1) {
            d2 = d1;
            d1 = x;
        } else if (x > d1 && x < d2) d2 = x;
        dfs(root.left);
        dfs(root.right);
    }

    public int findSecondMinimumValue(TreeNode root) {
        dfs(root);
        if (d2 == Long.MAX_VALUE) d2 = -1;
        return (int) d2;
    }

    public static void main(String[] args) {

        int nu = Integer.MIN_VALUE;  // 用Integer.MIN_VALUE表示二叉树 null
        int[] nums = new int[]{
                2,
                2, 5,
                nu, nu, 5, 7
        };
        TreeNode root = (new MyTree(nums)).getRoot();
        System.out.println((new Solution()).findSecondMinimumValue(root));
    }
}

File: Python/_0671_Second_Minimum_Node_In_a_Binary_Tree/Solution.py
# coding=utf-8
# Date: 2021/7/27 9:19
from MyTree import *


# 执行用时：28 ms, 在所有 Python3 提交中击败了95.70%的用户
# 内存消耗：14.9 MB, 在所有 Python3 提交中击败了42.58%的用户
class Solution:
    def __init__(self):
        self.d1, self.d2 = int(1e18), int(1e18)

    def findSecondMinimumValue(self, root: TreeNode) -> int:
        self.dfs(root)
        if self.d2 == int(1e18):
            self.d2 = -1
        return self.d2

    def dfs(self, root):
        if root is None:
            return
        x = root.val
        if x < self.d1:
            self.d2 = self.d1
            self.d1 = x
        elif self.d1 < x < self.d2:
            self.d2 = x
        self.dfs(root.left)
        self.dfs(root.right)


if __name__ == "__main__":
    nu = None
    nums = [
        2,
        2, 5,
        nu, nu, 5, 7
    ]
    root = MyTree(nums).getRoot()
    print(Solution().findSecondMinimumValue(root))  # 5
